File: Java/BrainTimer/app/src/main/java/com/example/braintimer/PlayingActivity.java
package com.example.braintimer;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.os.CountDownTimer;
import android.view.View;
import android.widget.TextView;

import com.example.braintimer.db.BrainTimerAppDb;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class PlayingActivity extends AppCompatActivity {

    TextView textViewTimer;
    TextView textViewQuestionCount;
    TextView textViewQuestion;
    TextView textViewAnswer1;
    TextView textViewAnswer2;
    TextView textViewAnswer3;
    TextView textViewAnswer4;
    SharedPreferences sharedPreferences;
    CountDownTimer timer;
    String username;
    int round = 0;
    int firstNumber = 0;
    int secondNumber = 0;
    int answer = 0;
    int questionCount = 0;
    BrainTimerAppDb db;

    @Override
    public void onBackPressed() {

        timer.cancel();
        long remainingTimer = Integer.parseInt(textViewTimer.getText().toString()) * 1000;

        AlertDialog.Builder dialog = new AlertDialog.Builder(this);
        dialog.setTitle("Warning");
        dialog.setMessage("Are you sure you want to exit the game ?");
        dialog.setNegativeButton("No", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                timer = initializeTimerAndReturn(remainingTimer,1000);
                timer.start();
            }
        });

        dialog.setPositiveButton("Yes", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                Intent intent = new Intent(getApplicationContext(),StartingActivity.class);
                startActivity(intent);
            }
        });

        dialog.show();
    }
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_playing);

        db = new BrainTimerAppDb(getApplicationContext());

        Intent intent = getIntent();
        username = intent.getStringExtra("name");

        textViewQuestionCount = findViewById(R.id.textViewQuestionCount);
        textViewQuestion = findViewById(R.id.textViewQuestion);
        textViewAnswer1 = findViewById(R.id.textViewAnswer1);
        textViewAnswer2 = findViewById(R.id.textViewAnswer2);
        textViewAnswer3 = findViewById(R.id.textViewAnswer3);
        textViewAnswer4 = findViewById(R.id.textViewAnswer4);
        textViewTimer = findViewById(R.id.textViewTimer);

        timer = initializeTimerAndReturn(30000,1000);

        timer.start();

        sharedPreferences = this.getSharedPreferences(getApplicationInfo().name, Context.MODE_PRIVATE);
        questionCount = sharedPreferences.getInt("questionCount",10);
        textViewQuestionCount.setText(String.format("%d/%d",round,questionCount));

        loadQuestion();
        loadAnswers();
    }
    private void loadQuestion(){
        Random rand = new Random();
        firstNumber = rand.nextInt(100) + 1;
        secondNumber = rand.nextInt(100) + 1;
        answer = firstNumber + secondNumber;

        textViewQuestion.setText(String.format("%d + %d",firstNumber,secondNumber));
    }
    private void selectAnswer(int selectedAnswer){

        timer.cancel();
        long remainingTimer = Integer.parseInt(textViewTimer.getText().toString()) * 1000;

        AlertDialog.Builder dialog = new AlertDialog.Builder(this);
        dialog.setTitle("Warning");
        dialog.setMessage("Are you sure you want to choose this answer ?");
        dialog.setNegativeButton("No", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                timer = initializeTimerAndReturn(remainingTimer,1000);
                timer.start();
            }
        });

        dialog.setPositiveButton("Yes", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                checkAnswer(selectedAnswer);
            }
        });

        dialog.show();

    }
    private void checkAnswer(int selectedAnswer){
        if(selectedAnswer != answer){

            db.insertScore(username,round,questionCount);

            Intent intent = new Intent(getApplicationContext(),StatusActivity.class);
            intent.putExtra("status","lose");
            intent.putExtra("correctQuestions",round);
            intent.putExtra("totalRounds",questionCount);
            startActivity(intent);
        }else if(round == questionCount){
            round++;
            Intent intent = new Intent(getApplicationContext(),StatusActivity.class);
            intent.putExtra("status","win");
            intent.putExtra("correctQuestions",questionCount);
            intent.putExtra("totalRounds",questionCount);
            startActivity(intent);
        }else{
            round++;

            loadQuestion();
            loadAnswers();
            textViewQuestionCount.setText(String.format("%d/%d",round,questionCount));
        }

    }
    private void loadAnswers(){
        Random rand = new Random();
        int firstAnswer = rand.nextInt(200) + 2;
        int secondAnswer = rand.nextInt(200) + 2;
        int thirdAnswer = rand.nextInt(200) + 2;
        int realAnswer = answer;

        TextView[] textViews = new TextView[]{textViewAnswer1,textViewAnswer2,textViewAnswer3,textViewAnswer4};
        int textViewCounter = 0;

        List<Integer> answers = new ArrayList<>();
        answers.add(firstAnswer);
        answers.add(secondAnswer);
        answers.add(thirdAnswer);
        answers.add(realAnswer);

        while(answers.size() != 0){
            Random answersRandom = new Random();
            int randomAnswerIndex = answersRandom.nextInt(answers.size());

            int answer = answers.get(randomAnswerIndex);
            textViews[textViewCounter].setText(Integer.toString(answer));

            textViews[textViewCounter].setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    selectAnswer(answer);
                }
            });
            textViewCounter++;
            answers.remove(randomAnswerIndex);
        }

    }
    private CountDownTimer initializeTimerAndReturn(long millisInFuture,long countDownInterval){
        return new CountDownTimer(millisInFuture,countDownInterval) {
            @Override
            public void onTick(long millisUntilFinished) {
                textViewTimer.setText(Long.toString(millisUntilFinished/1000));
            }

            @Override
            public void onFinish() {
                Intent intent = new Intent(getApplicationContext(),StatusActivity.class);
                intent.putExtra("status","lose");
                intent.putExtra("correctQuestions",round);
                intent.putExtra("totalRounds",questionCount);
                startActivity(intent);
            }
        };
    }
}

File: RockPaperScissorsApp/app/src/main/java/com/example/rockpaperscissorsapp/PlayingActivity.kt
package com.example.rockpaperscissorsapp

import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.activity.OnBackPressedCallback
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import kotlin.random.Random


class PlayingActivity : AppCompatActivity() {
    lateinit var imageViewComputerSelect : ImageView
    lateinit var imageViewYourSelect : ImageView
    lateinit var imageViewRockSelect : ImageView
    lateinit var imageViewPaperSelect : ImageView
    lateinit var imageViewScissorsSelect : ImageView

    lateinit var textViewWinStatus : TextView
    lateinit var textViewScore : TextView

    var yourSelection : Selection? = null

    var survivedRounds = 0
    var topScoreComputer = 0
    var topScorePlayer = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_playing)


        setCloseGameDialog()
        setElements()

        imageViewPaperSelect.setOnClickListener(View.OnClickListener {
            setImageBySelection(Selection.PAPER,imageViewYourSelect)
            yourSelection = Selection.PAPER
        })

        imageViewRockSelect.setOnClickListener(View.OnClickListener {
            setImageBySelection(Selection.ROCK,imageViewYourSelect)
            yourSelection = Selection.ROCK
        })

        imageViewScissorsSelect.setOnClickListener(View.OnClickListener {
            setImageBySelection(Selection.SCISSORS,imageViewYourSelect)
            yourSelection = Selection.SCISSORS
        })

        imageViewComputerSelect.setOnClickListener(View.OnClickListener {
            var alertDialog = AlertDialog.Builder(this)
            alertDialog.setMessage("Are you ready ?")
            alertDialog.setNegativeButton("No",null)
            alertDialog.setPositiveButton("Yes",DialogInterface.OnClickListener{ dialog, which ->
                startGame()
            })

            alertDialog.show()
        })
    }

    private fun setElements() {
        topScoreComputer = resources.getInteger(R.integer.topScoreComputer)
        topScorePlayer = resources.getInteger(R.integer.topScorePlayer)

        imageViewComputerSelect = findViewById(R.id.imageViewComputerSelect)
        imageViewYourSelect = findViewById(R.id.imageViewYourSelect)
        imageViewRockSelect = findViewById(R.id.imageViewRockSelect)
        imageViewPaperSelect = findViewById(R.id.imageViewPaperSelect)
        imageViewScissorsSelect = findViewById(R.id.imageViewScissorsSelect)

        textViewScore = findViewById(R.id.textViewScore)
        textViewWinStatus = findViewById(R.id.textViewWinStatus)
    }

    private fun setCloseGameDialog() {
        val callback: OnBackPressedCallback =
            object : OnBackPressedCallback(true) {
                override fun handleOnBackPressed() {
                    var alertDialog = AlertDialog.Builder(this@PlayingActivity)
                    alertDialog.setMessage("Are you sure you want to exit game ?")
                    alertDialog.setNegativeButton("No", null)
                    alertDialog.setPositiveButton(
                        "Yes",
                        DialogInterface.OnClickListener { dialog, which ->
                            finishAffinity()
                        })

                    alertDialog.show()
                }
            }

        this.onBackPressedDispatcher.addCallback(this, callback)
    }

    private fun startGame(){

        if(yourSelection == null){
            Toast.makeText(this,"You have to choose one of them",Toast.LENGTH_SHORT).show()
        }else{
            val randomNumber = Random.nextInt(0,3)
            val computerSelection = enumValues<Selection>()[randomNumber]
            setImageBySelection(computerSelection,imageViewComputerSelect)

            calculateWinStatus(yourSelection!!,computerSelection)

            survivedRounds++

            val score = Integer.parseInt(textViewScore.text.toString())
            if(score == topScoreComputer || score == topScorePlayer){
                calculateOverallStatus(score)
            }
        }

    }

    private fun setImageBySelection(selection:Selection,imageView:ImageView){
        when(selection){
            Selection.ROCK -> imageView.setImageResource(R.drawable.rock)
            Selection.PAPER -> imageView.setImageResource(R.drawable.paper)
            else -> imageView.setImageResource(R.drawable.scissors)
        }
    }

    private fun calculateWinStatus(yourSelection:Selection,computerSelection:Selection){
        when (yourSelection) {
            Selection.ROCK -> {
                when (computerSelection) {
                    Selection.PAPER -> {
                        textViewWinStatus.text = resources.getString(R.string.loseStatus)
                        textViewWinStatus.setTextColor(ContextCompat.getColor(this@PlayingActivity,R.color.red))
                        textViewScore.text = (Integer.parseInt(textViewScore.text.toString()) - 1).toString()
                    }
                    Selection.SCISSORS -> {
                        textViewWinStatus.text = resources.getString(R.string.winStatus)
                        textViewWinStatus.setTextColor(ContextCompat.getColor(this@PlayingActivity,R.color.green))
                        textViewScore.text = (Integer.parseInt(textViewScore.text.toString()) + 1).toString()
                    }
                    else -> {
                        textViewWinStatus.text = resources.getString(R.string.drawStatus)
                        textViewWinStatus.setTextColor(ContextCompat.getColor(this@PlayingActivity,R.color.yellow))
                    }
                }
            }
            Selection.PAPER -> {
                when (computerSelection) {
                    Selection.SCISSORS -> {
                        textViewWinStatus.text = resources.getString(R.string.loseStatus)
                        textViewWinStatus.setTextColor(ContextCompat.getColor(this@PlayingActivity,R.color.red))
                        textViewScore.text = (Integer.parseInt(textViewScore.text.toString()) - 1).toString()
                    }
                    Selection.ROCK -> {
                        textViewWinStatus.text = resources.getString(R.string.winStatus)
                        textViewWinStatus.setTextColor(ContextCompat.getColor(this@PlayingActivity,R.color.green))
                        textViewScore.text = (Integer.parseInt(textViewScore.text.toString()) + 1).toString()
                    }
                    else -> {
                        textViewWinStatus.text = resources.getString(R.string.drawStatus)
                        textViewWinStatus.setTextColor(ContextCompat.getColor(this@PlayingActivity,R.color.yellow))
                    }
                }
            }
            else -> {
                when (computerSelection) {
                    Selection.ROCK -> {
                        textViewWinStatus.text = resources.getString(R.string.loseStatus)
                        textViewWinStatus.setTextColor(ContextCompat.getColor(this@PlayingActivity,R.color.red))
                        textViewScore.text = (Integer.parseInt(textViewScore.text.toString()) - 1).toString()
                    }
                    Selection.PAPER -> {
                        textViewWinStatus.text = resources.getString(R.string.winStatus)
                        textViewWinStatus.setTextColor(ContextCompat.getColor(this@PlayingActivity,R.color.green))
                        textViewScore.text = (Integer.parseInt(textViewScore.text.toString()) + 1).toString()
                    }
                    else -> {
                        textViewWinStatus.text = resources.getString(R.string.drawStatus)
                        textViewWinStatus.setTextColor(ContextCompat.getColor(this@PlayingActivity,R.color.yellow))
                    }
                }
            }
        }
    }

    private fun calculateOverallStatus(score:Int){
        val intent = Intent(this,EndingActivity::class.java)
        if(score == resources.getInteger(R.integer.topScorePlayer)){
            intent.putExtra("survivedRounds",survivedRounds)
            intent.putExtra("winStatus",resources.getString(R.string.winStatus))
        }else if(score == resources.getInteger(R.integer.topScoreComputer)){
            intent.putExtra("survivedRounds",survivedRounds)
            intent.putExtra("winStatus",resources.getString(R.string.loseStatus))
        }

        startActivity(intent)
    }


}